cache/cache.go

```diff
@@ -0,0 +1,928 @@
+	f.totalWorkers = f.OriginalSettingWorkers()
+	f.chunkMemory = !f.OriginalSettingChunkNoMemory()
+	f.warmUp = false
+}
+
+// NewObject finds the Object at remote.
+func (f *Fs) NewObject(remote string) (fs.Object, error) {
+	co := NewObject(f, remote)
+	err := f.cache.GetObject(co)
+	if err == nil {
+		return co, nil
+	}
+	obj, err := f.Fs.NewObject(remote)
+	if err != nil {
+		return nil, err
+	}
+	co = ObjectFromOriginal(f, obj)
+	co.Persist()
+	return co, nil
+}
+
+// List the objects and directories in dir into entries
+func (f *Fs) List(dir string) (entries fs.DirEntries, err error) {
+	// clean cache
+	go f.CleanUpCache(false)
+
+	cd := NewDirectory(f, dir)
+	entries, err = f.Cache().GetDirEntries(cd)
+	if err != nil {
+		fs.Debugf(dir, "no dir entries in cache: %v", err)
+	} else if len(entries) == 0 {
+		// TODO: read empty dirs from source?
+	} else {
+		return entries, nil
+	}
+
+	entries, err = f.Fs.List(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var cachedEntries fs.DirEntries
+	for _, entry := range entries {
+		switch o := entry.(type) {
+		case fs.Object:
+			co := ObjectFromOriginal(f, o)
+			co.Persist()
+			cachedEntries = append(cachedEntries, co)
+		case fs.Directory:
+			cd := DirectoryFromOriginal(f, o)
+			err = f.Cache().AddDir(cd)
+			cachedEntries = append(cachedEntries, cd)
+		default:
+			err = errors.Errorf("Unknown object type %T", entry)
+		}
+	}
+	if err != nil {
+		fs.Errorf(dir, "err caching listing: %v", err)
+	}
+
+	return cachedEntries, nil
+}
+
+func (f *Fs) recurse(dir string, list *fs.ListRHelper) error {
+	entries, err := f.List(dir)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(entries); i++ {
+		innerDir, ok := entries[i].(fs.Directory)
+		if ok {
+			err := f.recurse(innerDir.Remote(), list)
+			if err != nil {
+				return err
+			}
+		}
+
+		err := list.Add(entries[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ListR lists the objects and directories of the Fs starting
+// from dir recursively into out.
+func (f *Fs) ListR(dir string, callback fs.ListRCallback) (err error) {
+	fs.Debugf(f, "list recursively from '%s'", dir)
+
+	// we check if the source FS supports ListR
+	// if it does, we'll use that to get all the entries, cache them and return
+	do := f.Fs.Features().ListR
+	if do != nil {
+		return f.Fs.Features().ListR(dir, func(entries fs.DirEntries) error {
+			// we got called back with a set of entries so let's cache them and call the original callback
+			for _, entry := range entries {
+				switch o := entry.(type) {
+				case fs.Object:
+					_ = f.Cache().AddObject(ObjectFromOriginal(f, o))
+				case fs.Directory:
+					_ = f.Cache().AddDir(DirectoryFromOriginal(f, o))
+				default:
+					return errors.Errorf("Unknown object type %T", entry)
+				}
+			}
+
+			// call the original callback
+			return callback(entries)
+		})
+	}
+
+	// if we're here, we're gonna do a standard recursive traversal and cache everything
```

Comment:
It's not exactly like that. Here's how I imagined this feature:

1. if the wrapped remote supports `ListR` then I use it and add my callback that caches every batch received and then calls the original callback which was passed to the cache `ListR` - there's a return to this `if` block. It will return here if enters
2. if it's not supported - I do a recursive listing using cache's `List` (caching every entry and dir)

Reply:
You are right - I mis-read that bit of code!

You don't need to re-implement ListR if the underlying remote doesn't support it.  The f.Mask() will mask it out and rclone will do essentially exactly what your `recurse` function does.